zoneinfo/src/zone.rs

```diff
@@ -0,0 +1,506 @@
+            self.year_seconds = match self.start_kind {
+                QualifiedTimeKind::Universal => self.epoch_year,
+                QualifiedTimeKind::Standard => self.epoch_year - zone.std_offset.as_secs(),
+                // Uh, how to handle dst. Does it matter? This will prob blow up on southern hemisphere
+                QualifiedTimeKind::Local => {
+                    self.epoch_year - zone.std_offset.as_secs() - self.saving.as_secs()
+                }
+            };
+        }
+    }
+
+    pub(crate) fn is_zone_beyond_year(&self) -> bool {
+        self.year_seconds < self.use_start && !self.is_start_in_year_range()
+    }
+
+    pub(crate) fn in_zone_skippable(&self, until_time: i64) -> bool {
+        !(self.use_start..=until_time).contains(&self.year_seconds)
+            && !self.is_start_in_year_range()
+    }
+
+    pub(crate) fn is_start_in_year_range(&self) -> bool {
+        self.year_range.contains(&self.use_start)
+    }
+
+    pub(crate) fn zone_was_rule(&self) -> bool {
+        matches!(self.previous_rule, RuleIdentifier::Rule(_))
+    }
+}
+#[derive(Debug, Clone, Default)]
+pub struct ZoneTable {
+    pub table: Vec<ZoneEntry>,
+    pub associates: HashMap<String, RuleTable>,
+}
+
+impl IntoIterator for ZoneTable {
+    type Item = ZoneEntry;
+    type IntoIter = alloc::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.table.into_iter()
+    }
+}
+
+impl ZoneTable {
+    pub fn associate_rules(&mut self, rules: &HashMap<String, RuleTable>) {
+        for entry in &mut self.table {
+            if let RuleIdentifier::Rule(associate_rule) = &entry.rule {
+                if self.associates.contains_key(associate_rule) {
+                    continue;
+                }
+                if let Some(rules) = rules.get(associate_rule).cloned() {
+                    let _ = self.associates.insert(associate_rule.clone(), rules);
+                }
+            }
+        }
+    }
+
+    /// Get teh first transition time for this zone table.
+    ///
+    /// No transition will be lower than this.
+    fn get_first_transition_time(&self) -> i64 {
+        let lmt_entry = &self.table[0];
+        let lmt_date = lmt_entry.date.expect("must exist");
+        lmt_date.as_precise_ut_time(&lmt_entry.std_offset, &Time::default())
+    }
+
+    pub fn get_first_transition(&self) -> Option<Transition> {
+        let lmt_entry = &self.table[0];
+        lmt_entry.date.map(|date| {
+            Transition {
+                at_time: self.get_first_transition_time(),
+                offset: lmt_entry.std_offset.as_secs(),
+                dst: false,
+                letter: None,
+                time_type: date.time.time_kind(),
+                format: String::from("LMT"), // This value is constant for the first transition
+            }
+        })
+    }
+
+    pub(crate) fn calculate_transitions_for_year(
```

Comment:
This method is were a large bulk of the zoneinfo compilation work lives.

I tried my best to document it because it's basically a rough reverse engineering of [`outzone`](https://github.com/eggert/tz/blob/main/zic.c#L3149).